eth/tracers/tracker.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package tracers
+
+import (
+	"fmt"
+	"sync"
+)
+
+// stateTracker is an auxiliary tool used to cache the release functions of all
+// used trace states, and to determine whether the creation of trace state needs
+// to be paused in case there are too many states waiting for tracing.
+type stateTracker struct {
+	limit    int                // Maximum number of states allowed waiting for tracing
+	oldest   uint64             // The number of the oldest state which is still using for trace
+	used     []bool             // List of flags indicating whether the trace state has been used up
+	releases []StateReleaseFunc // List of trace state release functions waiting to be called
+	cond     *sync.Cond
+	lock     *sync.RWMutex
+}
+
+// newStateTracker initializes the tracker with provided state limits and
+// the number of the first state that will be used.
+func newStateTracker(limit int, oldest uint64) *stateTracker {
+	lock := new(sync.RWMutex)
+	return &stateTracker{
+		limit:  limit,
+		oldest: oldest,
+		used:   make([]bool, limit),
+		cond:   sync.NewCond(lock),
+		lock:   lock,
+	}
+}
+
+// releaseState marks the state specified by the number as released and caches
+// the corresponding release functions internally.
+func (t *stateTracker) releaseState(number uint64, release StateReleaseFunc) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	// Set the state as used, the corresponding flag is indexed by
+	// the distance between the specified state and the oldest state
+	// which is still using for trace.
+	t.used[int(number-t.oldest)] = true
+
+	// If the oldest state is used up, update the oldest marker by moving
+	// it to the next state which is not used up.
+	if number == t.oldest {
+		var count int
+		for i := 0; i < len(t.used); i++ {
+			if !t.used[i] {
+				break
+			}
+			count += 1
+		}
+		t.oldest += uint64(count)
+		copy(t.used, t.used[count:])
+
+		// The tail of t.used is left unchanged. It will be overwritten
+		// by the following states anyway.
```

Comment:
This is not obvious to me. Perhaps you can clarify what you mean by "following states" ?

Reply:
For example, the size of used array is 5, and the current state os this array is [true, true, false, true, false]
after the loop, we find that the first two flags are marked as true, and we left-shift the array by 2, it's [false, true, false, true, false]

You can see that the last two elements are left unchanged. And actually they are useless dirty data. I didn't explicitly clean them up since although they are wrong dirty data, but they don't belong to any state. 

Whenever we set the flag for the next state, we will overwrite the value of these elements.

Reply:
How about another example... 

For example, the size of used array is 5, and the current state os this array is [true, true, false, true, true]
after the loop, we find that the first two flags are marked as true, and we left-shift the array by 2, it's [false, true, true, true, true]

You can see that the last two elements are left unchanged (`true`). And actually they are useless dirty data. 


Now, what I'm curious about is what happens when  the first `false` is turned into `true`. Suddenly all states are flushed as done?

Reply:
... To clarify: from that point onward, nothing will ever become `false` inside the `t.used` slice again. As far as I can tell. And that doesn't seem right.

Reply:
Hm. I think the dissonance has been that I've seen this as data type which 'moves' along with the state generation. But, in fact, it's meant to be used once? 
So you init it with the full range that you expect to ever need? And then you never extend it beyond that? 

Because if so, wouldn't be better to.. 

- Instead of moving subslices to the left, in `t.used`, 
- keep one single index of the first non-true element is

Reply:
Nope, strike that. I still believe there's something wrong here. At least I can't quite make sense of how this works, if it works :)